src/jrd/vio.cpp

```diff
@@ -4556,12 +4557,15 @@ bool VIO_writelock(thread_db* tdbb, record_param* org_rpb, jrd_tra* transaction)
 				}
 			}
 			org_rpb->rpb_runtime_flags |= RPB_refetch;
-			return false;
+			return WriteLockResult::CONFLICTED;
```

Comment:
I'm wondering whether we could return `WriteLockResult::SKIPPED` for _ReadCommitted_ transactions for the `PREPARE_DELETE` case, thus avoiding a refetch inside `LockedStream::LockRecord()` (or request restart in the _ReadConsistency_ mode). @hvlad, what do you think? Do I miss anything?

Reply:
I need some time to review, will report later.

Reply:
Looks like you're correct here

Reply:
Do you mean that or something different?

```diff
diff --git a/src/jrd/vio.cpp b/src/jrd/vio.cpp
index 9b7cf4679c..d86336ae43 100644
--- a/src/jrd/vio.cpp
+++ b/src/jrd/vio.cpp
@@ -4545,8 +4545,12 @@ WriteLockResult VIO_writelock(thread_db* tdbb, record_param* org_rpb, jrd_tra* t
        switch (prepare_update(tdbb, transaction, org_rpb->rpb_transaction_nr, org_rpb, &temp, &new_rpb,
                                                   stack, TriState(skipLocked)))
        {
-               case PrepareResult::CONFLICT:
                case PrepareResult::DELETED:
+                       if (skipLocked && (transaction->tra_flags & TRA_read_committed))
+                               return WriteLockResult::SKIPPED;
+                       // fall thru
+
+               case PrepareResult::CONFLICT:
                        if ((transaction->tra_flags & TRA_read_consistency))
                        {
                                Request* top_request = tdbb->getRequest()->req_snapshot.m_owner;
```

Reply:
Yes, something like this.